### PR TITLE
Vojtakai/convex finance staking

### PR DESCRIPTION
**Pull request by VojtaKai**
# Description

Implementation of a staking bridge for Convex Finance. You can stake and withdraw your Curve LP Tokens anytime.
Each Curve pool has its own LP Tokens and you can get them as a reward for lending your assets to their pool. These LP Tokens can be used further for staking. In this case, with Convex Finance. Convex Finance mints its own Convex Tokens that are later staked. Convex Tokens are not returned. For the Rollup Processor to be able to track these staked assets, the bridge mints its own Tokens - CSB - which ownership is then given to the Rollup Processor.

# Checklist:

- [x] I have reviewed my diff in github, line by line.
- [x] Every change is related to the PR description.
- [x] There are no unexpected formatting changes, or superfluous debug logs.
- [ ] The branch has been rebased against the head of its merge target.
- [ ] I'm happy for the PR to be merged at the reviewers next convenience.
- [x] NatSpec documentation of all the non-test functions is present and is complete.
- [x] Continuous integration (CI) passes.
- [x] Command `forge coverage --match-contract MyContract` returns 100% line coverage.
- [x] All the possible reverts are tested.

### CODE REVIEW

File: src/bridges/convex/ConvexStakingBridge.sol [hunk 1]
```diff
@@ -0,0 +1,195 @@
...
+    constructor(address _rollupProcessor) BridgeBase(_rollupProcessor) {
+        addInvalidPoolPid(48);
+        // _mint(address(this), DUST) // for optimization
+    }
+
+    /** 
+    @notice Add a new pool pid that should not be allowed to interact with.
+    @param poolPid Id of a pool to add.
+    */
+    function addInvalidPoolPid(uint poolPid) public onlyRollup {
```
Comment (benesjan) [inline src/bridges/convex/ConvexStakingBridge.sol:70]: The bridge doesn't manage to get deployed in the E2E test because you use the onlyRollup modifier in this function which gets called from within constructor and reverts. In the unit tests, the test contract itself is passed to the bridge as the rollupProcessor and for this reason the deployment succeeds there.

File: src/bridges/convex/ConvexStakingBridge.sol [hunk 2]
```diff
@@ -0,0 +1,258 @@
...
+        if (
+            _inputAssetA.assetType == AztecTypes.AztecAssetType.ERC20 &&
+            _outputAssetA.assetType == AztecTypes.AztecAssetType.VIRTUAL
+        ) {
+            // deposit
+            if (!pools[_inputAssetA.erc20Address].exists) {
+                revert ErrorLib.InvalidInputA();
+            }
+
+            selectedPool = pools[_inputAssetA.erc20Address];
```
Comment (benesjan) [inline src/bridges/convex/ConvexStakingBridge.sol:141]: Move this line above line 137 and on line 137 use selectedPool instead. That will save multiple SLOADS and save gas since the struct info is in multiple storage slots.

File: src/bridges/convex/ConvexStakingBridge.sol [hunk 3]
```diff
@@ -0,0 +1,258 @@
...
+contract ConvexStakingBridge is BridgeBase {
+    using SafeERC20 for IConvexBooster;
+    using SafeERC20 for ICurveLpToken;
+
+    struct PoolInfo {
+        uint256 poolId;
+        address curveLpToken;
+        address convexToken;
+        address curveRewards;
+        bool exists;
```
Comment (benesjan) [inline src/bridges/convex/ConvexStakingBridge.sol:40]: The `exists` value is a bit redundant since you could simply check if any of the struct's addresses are equal to `address(0)`. In this case removing it won't save that much gas since the value will get packed with `curveRewards` address but I would remove it anyway.

File: src/bridges/convex/ConvexStakingBridge.sol [hunk 4]
```diff
@@ -0,0 +1,258 @@
...
+        } else if (
+            _inputAssetA.assetType == AztecTypes.AztecAssetType.VIRTUAL &&
+            _outputAssetA.assetType == AztecTypes.AztecAssetType.ERC20
+        ) {
+            // withdraw
+            if (!interactions[_inputAssetA.id].exists) {
+                revert UnknownVirtualAsset();
+            }
+            // withdraw amount needs to be equal to the staked value of the interaction
+            if (interactions[_inputAssetA.id].valueStaked != _totalInputValue) {
```
Comment (benesjan) [inline src/bridges/convex/ConvexStakingBridge.sol:166]: This is not feasible because you can't ensure that all the users which were present in the interaction would exit at the same time. I realized that unfortunately the virtual assets were not a good idea here because they would currently break aggregation --> virtual asset id is set to `interactionNonce` which makes it impossible to aggregate users holding virtualAssets corresponding to the same pool when interactionNonce is different. Doing the change should be fairly straighforward though. In the `_loadPools(...)` function deploy a new ERC20 for each pool and in convert function be minting and burning the token. You want to deploy mintable ERC20 where the bridge will have the rights to mint and burn. Check out Openzeppelin's ERC20 token implementations.

**Review by benesjan — CHANGES_REQUESTED** [review]

File: src/bridges/convex/ConvexStakingBridge.sol [hunk 5]
```diff
@@ -0,0 +1,258 @@
...
+    receive() external payable {}
+
+    /**
+    @notice Stake and unstake Curve LP tokens through Convex Finance Booster anytime.
+    @notice Convert rate between Curve LP token and corresponding Convex LP token is 1:1
+    @notice Stake == Deposit, Unstake == Withdraw
+    @param _inputAssetA Curve LP token (staking), virtual asset (unstaking)
+    @param _outputAssetA Virtual asset (staking), Curve LP token (unstaking)
+    @param _totalInputValue Total number of Curve LP tokens to deposit / withdraw
+    @param _auxData Data to claim staking rewards
```
[the remark this answers is not in the document]
Comment (VojtaKai) [reply src/bridges/convex/ConvexStakingBridge.sol:100]: Resolved

File: src/bridges/convex/ConvexStakingBridge.sol [hunk 6]
```diff
@@ -0,0 +1,258 @@
...
+
+import {SafeERC20} from "@openzeppelin/contracts/token/ERC20/utils/SafeERC20.sol";
+import {ERC20} from "@openzeppelin/contracts/token/ERC20/ERC20.sol";
+import {IERC20} from "@openzeppelin/contracts/token/ERC20/IERC20.sol";
+
+import {BridgeBase} from "../base/BridgeBase.sol";
+import {AztecTypes} from "rollup-encoder/libraries/AztecTypes.sol";
+import {ICurvePool} from "../../interfaces/curve/ICurvePool.sol";
+import {ErrorLib} from "../base/ErrorLib.sol";
+import {IRollupProcessor} from "rollup-encoder/interfaces/IRollupProcessor.sol";
```
[the remark this answers is not in the document]
Comment (VojtaKai) [reply src/bridges/convex/ConvexStakingBridge.sol:13]: Resolved

File: src/bridges/convex/ConvexStakingBridge.sol [hunk 7]
```diff
@@ -0,0 +1,258 @@
...
+    error IncorrectInteractionValue(uint256 stakedValue, uint256 valueToWithdraw);
+
+    constructor(address _rollupProcessor) BridgeBase(_rollupProcessor) {
+        _loadPools();
+
+        uint256[] memory criterias = new uint256[](2 * lastPoolLength);
+        uint32[] memory gasUsage = new uint32[](2 * lastPoolLength);
+        uint32[] memory minGasPerMinute = new uint32[](2 * lastPoolLength);
+
+        for (uint256 i = 0; i < lastPoolLength; i++) {
```
[the remark this answers is not in the document]
Comment (VojtaKai) [reply src/bridges/convex/ConvexStakingBridge.sol:74]: Resolved.

File: src/bridges/convex/ConvexStakingBridge.sol [hunk 8]
```diff
@@ -0,0 +1,258 @@
...
+            uint256 outputValueA,
+            uint256,
+            bool
+        )
+    {
+        if (_totalInputValue == 0) {
+            revert ErrorLib.InvalidInputAmount();
+        }
+
+        _loadPools();
```
[the remark this answers is not in the document]
Comment (VojtaKai) [reply src/bridges/convex/ConvexStakingBridge.sol:128]: Resolved

File: src/bridges/convex/ConvexStakingBridge.sol [hunk 9]
```diff
@@ -0,0 +1,258 @@
...
+        bool exists;
+    }
+
+    // Contracts
+    IConvexBooster public constant BOOSTER = IConvexBooster(0xF403C135812408BFbE8713b5A23a04b3D48AAE31);
+    ICurveLpToken public curveLpToken;
+    ICurveRewards public curveRewards;
+
+    // Pools
+    uint256 public lastPoolLength;
```
[the remark this answers is not in the document]
Comment (VojtaKai) [reply src/bridges/convex/ConvexStakingBridge.sol:55]: Resolved

File: src/bridges/convex/ConvexStakingBridge.sol [hunk 10]
```diff
@@ -0,0 +1,258 @@
...
+ The receipt is then used to withdraw the staked means when a virtual asset of the matching ID is provided on input. 
+ @dev Synchronous and stateful bridge that keeps track of deposit interactions with the bridge that are later used for withdrawal.
+ @author Vojtech Kaiser
+ */
+contract ConvexStakingBridge is BridgeBase {
+    using SafeERC20 for IConvexBooster;
+    using SafeERC20 for ICurveLpToken;
+
+    struct PoolInfo {
+        uint256 poolId;
```
[the remark this answers is not in the document]
Comment (VojtaKai) [reply src/bridges/convex/ConvexStakingBridge.sol:36]: Resolved

File: src/bridges/convex/ConvexStakingBridge.sol [hunk 11]
```diff
@@ -0,0 +1,258 @@
...
+                pools[curveLpToken] = PoolInfo(i, curveLpToken, convexToken, curveRewards, true);
+            }
+            lastPoolLength = currentPoolLength;
+        }
+    }
+
+    /** 
+    @notice Sets up pool specific tokens.
+    */
+    function _setTokens(PoolInfo memory _selectedPool) internal {
```
[the remark this answers is not in the document]
Comment (VojtaKai) [reply src/bridges/convex/ConvexStakingBridge.sol:238]: Resolved

File: src/bridges/convex/ConvexStakingBridge.sol (hunk 2 above)
Comment (VojtaKai) [reply src/bridges/convex/ConvexStakingBridge.sol:141]: Resolved

File: src/bridges/convex/ConvexStakingBridge.sol (hunk 3 above)
Comment (VojtaKai) [reply src/bridges/convex/ConvexStakingBridge.sol:40]: Resolved

File: src/bridges/convex/ConvexStakingBridge.sol [hunk 12]
```diff
@@ -0,0 +1,258 @@
...
+ @dev Synchronous and stateful bridge that keeps track of deposit interactions with the bridge that are later used for withdrawal.
+ @author Vojtech Kaiser
+ */
+contract ConvexStakingBridge is BridgeBase {
+    using SafeERC20 for IConvexBooster;
+    using SafeERC20 for ICurveLpToken;
+
+    struct PoolInfo {
+        uint256 poolId;
+        address curveLpToken;
```
[the remark this answers is not in the document]
Comment (VojtaKai) [reply src/bridges/convex/ConvexStakingBridge.sol:37]: Resolved

File: src/bridges/convex/ConvexStakingBridge.sol [hunk 13]
```diff
@@ -0,0 +1,258 @@
...
+            if (!pools[_inputAssetA.erc20Address].exists) {
+                revert ErrorLib.InvalidInputA();
+            }
+
+            selectedPool = pools[_inputAssetA.erc20Address];
+
+            _setTokens(selectedPool);
+
+            // approvals
+            curveLpToken.approve(address(BOOSTER), _totalInputValue);
```
[the remark this answers is not in the document]
Comment (VojtaKai) [reply src/bridges/convex/ConvexStakingBridge.sol:146]: Resolved

File: src/bridges/convex/ConvexStakingBridge.sol [hunk 14]
```diff
@@ -0,0 +1,223 @@
...
+     * @return The criteria
+     */
+    function computeCriteria(
+        AztecTypes.AztecAsset calldata _inputAssetA,
+        AztecTypes.AztecAsset calldata,
+        AztecTypes.AztecAsset calldata _outputAssetA,
+        AztecTypes.AztecAsset calldata,
+        uint64
+    ) public view override(BridgeBase) returns (uint256) {
+        if (deployedTokens[_inputAssetA.erc20Address] == _outputAssetA.erc20Address) {
```
Comment (benesjan) [inline src/bridges/convex/ConvexStakingBridge.sol:209]: Both branches are computing the same thing.

File: src/bridges/convex/ConvexStakingBridge.sol [hunk 15]
```diff
@@ -0,0 +1,223 @@
...
+        uint96 poolId;
+        address convexToken;
+        address curveRewards;
+    }
+
+    // Convex Finance Booster
+    IConvexBooster public constant BOOSTER = IConvexBooster(0xF403C135812408BFbE8713b5A23a04b3D48AAE31);
+
+    // Pools
+    uint256 public poolLength;
```
Comment (benesjan) [inline src/bridges/convex/ConvexStakingBridge.sol:48]: Please call it poolsLength since it references pools mapping.

File: src/bridges/convex/ConvexStakingBridge.sol [hunk 16]
```diff
@@ -0,0 +1,223 @@
...
+            _rollupBeneficiary
+        );
+    }
+
+    /**
+    @notice Loads pool information for all pools supported by Convex Finance.
+    @notice Set allowance for Rollup's Curve LP tokens.
+    @notice Cached. Loads only new pools.
+    */
+    function loadPools() public {
```
Comment (benesjan) [inline src/bridges/convex/ConvexStakingBridge.sol:176]: Given that we have to list all the tokens on RollupProcessor as well it makes sense to change this functions so that it takes `poolId` as a parameter and sets up only that one pool. Deploying contracts is expensive so it will save us a lot of money on gas on the pools we would not use. I propose having this function signature:
`function listPool(uint256 _poolId) external;`.

File: src/bridges/convex/ConvexStakingBridge.sol [hunk 17]
```diff
@@ -0,0 +1,223 @@
...
+        if (
+            _inputAssetA.assetType != AztecTypes.AztecAssetType.ERC20 ||
+            _outputAssetA.assetType != AztecTypes.AztecAssetType.ERC20
+        ) {
+            revert InvalidAssetType();
+        }
+
+        if (
+            deployedTokens[_inputAssetA.erc20Address] == address(0) &&
+            deployedTokens[_outputAssetA.erc20Address] == address(0)
```
Comment (benesjan) [inline src/bridges/convex/ConvexStakingBridge.sol:124]: You are accessing storage variables on the 2 lines above and then again on line 131 and 143. Accessing storage is expensive so pls pre-load it to local variables and then do the checks on those.

File: src/bridges/convex/ConvexStakingBridge.sol [hunk 18]
```diff
@@ -0,0 +1,223 @@
...
+
+    constructor(address _rollupProcessor) BridgeBase(_rollupProcessor) {
+        loadPools();
+
+        uint256[] memory criterias = new uint256[](poolLength);
+        uint32[] memory gasUsage = new uint32[](poolLength);
+        uint32[] memory minGasPerMinute = new uint32[](poolLength);
+
+        for (uint256 i = 0; i < poolLength; i++) {
+            (address curveLpToken, , , , , ) = BOOSTER.poolInfo(i);
```
Comment (benesjan) [inline src/bridges/convex/ConvexStakingBridge.sol:67]: Move this to the `loadPool` function since there you are looping over the same data as well. Also you would not have the subsidy criteria set for pools which are added after the deployment of this contract.

File: src/bridges/convex/ConvexStakingBridge.sol [hunk 19]
```diff
@@ -0,0 +1,223 @@
...
+        payable
+        override(BridgeBase)
+        onlyRollup
+        returns (
+            uint256 outputValueA,
+            uint256,
+            bool
+        )
+    {
+        if (_totalInputValue == 0) {
```
Comment (benesjan) [inline src/bridges/convex/ConvexStakingBridge.sol:111]: Our `RollupProcessor` contract doesn't allow the value to be 0 so the check is redundant.

File: src/bridges/convex/ConvexStakingBridge.sol [hunk 20]
```diff
@@ -0,0 +1,223 @@
...
+    */
+    function loadPools() public {
+        uint256 currentPoolLength = BOOSTER.poolLength();
+        if (currentPoolLength != poolLength) {
+            for (uint256 i = poolLength; i < currentPoolLength; i++) {
+                (address curveLpToken, address convexToken, , address curveRewards, , ) = BOOSTER.poolInfo(i);
+                pools[curveLpToken] = PoolInfo(uint96(i), convexToken, curveRewards);
+
+                // deploy token, log token address
+                address deployedToken = address(new RepresentingConvexToken("RepresentingConvexToken", "RCT"));
```
Comment (benesjan) [inline src/bridges/convex/ConvexStakingBridge.sol:184]: Since deploying this contract is expensive it would make sense to optimize it. One way to do it is to use a proxy pattern. Deploy this contract only in constructor and save it's address to implementation immutable variable. Then here use the [clone library](https://github.com/OpenZeppelin/openzeppelin-contracts/blob/master/contracts/proxy/Clones.sol) to deploy a small proxy which will point to the implementation.

File: src/bridges/convex/ConvexStakingBridge.sol [hunk 21]
```diff
@@ -0,0 +1,223 @@
...
+    ) public view override(BridgeBase) returns (uint256) {
+        if (deployedTokens[_inputAssetA.erc20Address] == _outputAssetA.erc20Address) {
+            return uint256(keccak256(abi.encodePacked(_inputAssetA.erc20Address, _outputAssetA.erc20Address)));
+        } else {
+            return uint256(keccak256(abi.encodePacked(_inputAssetA.erc20Address, _outputAssetA.erc20Address)));
+        }
+    }
+}
+
+contract RepresentingConvexToken is ERC20, ERC20Burnable, Ownable {
```
Comment (benesjan) [inline src/bridges/convex/ConvexStakingBridge.sol:217]: Since `ERC20Burnable` extends `ERC20` extending eRC20 here is redundant.

**Review by benesjan — CHANGES_REQUESTED** [review]

File: src/bridges/convex/ConvexStakingBridge.sol (hunk 14 above)
Comment (VojtaKai) [reply src/bridges/convex/ConvexStakingBridge.sol:209]: Resolved

File: src/bridges/convex/ConvexStakingBridge.sol (hunk 15 above)
Comment (VojtaKai) [reply src/bridges/convex/ConvexStakingBridge.sol:48]: Resolved

File: src/bridges/convex/ConvexStakingBridge.sol (hunk 16 above)
Comment (VojtaKai) [reply src/bridges/convex/ConvexStakingBridge.sol:176]: Resolved

File: src/bridges/convex/ConvexStakingBridge.sol (hunk 17 above)
Comment (VojtaKai) [reply src/bridges/convex/ConvexStakingBridge.sol:124]: Resolved

File: src/bridges/convex/ConvexStakingBridge.sol (hunk 18 above)
Comment (VojtaKai) [reply src/bridges/convex/ConvexStakingBridge.sol:67]: Resolved

File: src/bridges/convex/ConvexStakingBridge.sol (hunk 19 above)
Comment (VojtaKai) [reply src/bridges/convex/ConvexStakingBridge.sol:111]: Resolved

File: src/bridges/convex/ConvexStakingBridge.sol (hunk 21 above)
Comment (VojtaKai) [reply src/bridges/convex/ConvexStakingBridge.sol:217]: Resolved.
Question: Does it somehow (negatively) impact the performance? Or it is just not necessary since we override a single function?

File: src/bridges/convex/ConvexStakingBridge.sol (hunk 20 above)
Comment (VojtaKai) [reply src/bridges/convex/ConvexStakingBridge.sol:184]: Resolved

File: src/bridges/convex/ConvexStakingBridge.sol [hunk 22]
```diff
@@ -0,0 +1,265 @@
...
+ RCT ERC20 token is deployed for each loaded pool and mirrors balance of minted Convex LP tokens for the bridge.
+ The difference is that RCT is owned directly by the bridge whereas Convex LP tokens are not and cannot be transferred to the Rollup Processor.
+ @dev Synchronous and stateless bridge
+ @author Vojtech Kaiser
+ */
+contract ConvexStakingBridge is BridgeBase {
+    using SafeERC20 for IConvexBooster;
+    using SafeERC20 for ICurveLpToken;
+
+    struct PoolInfo {
```
Comment (benesjan) [inline src/bridges/convex/ConvexStakingBridge.sol:34]: Please describe params of this struct using the NatSpec format.

**Review by benesjan — CHANGES_REQUESTED** [review]

File: src/bridges/convex/ConvexStakingBridge.sol [hunk 23]
```diff
@@ -0,0 +1,265 @@
...
+     * @return The criteria
+     */
+    function computeCriteria(
+        AztecTypes.AztecAsset calldata _inputAssetA,
+        AztecTypes.AztecAsset calldata,
+        AztecTypes.AztecAsset calldata _outputAssetA,
+        AztecTypes.AztecAsset calldata,
+        uint64
+    ) public view override(BridgeBase) returns (uint256) {
+        if (deployedClones[_inputAssetA.erc20Address] == _outputAssetA.erc20Address) {
```
[the remark this answers is not in the document]
Comment (VojtaKai) [reply src/bridges/convex/ConvexStakingBridge.sol:194]: Yes

File: src/bridges/convex/ConvexStakingBridge.sol [hunk 24]
```diff
@@ -0,0 +1,265 @@
...
+            revert ErrorLib.InvalidOutputA();
+        }
+
+        _claimSubsidy(SUBSIDY, _inputAssetA, _inputAssetB, _outputAssetA, _outputAssetB, _rollupBeneficiary);
+    }
+
+    /**
+    @notice Internal function to pay out subsidy to the rollupBeneficiary
+    */
+    function _claimSubsidy(
```
[the remark this answers is not in the document]
Comment (VojtaKai) [reply src/bridges/convex/ConvexStakingBridge.sol:124]: Convert complained about too many local variables - stack too deep so I decided to move this one out.

File: src/bridges/convex/ConvexStakingBridge.sol (hunk 23 above)
[the remark this answers is not in the document]
Comment (benesjan) [reply src/bridges/convex/ConvexStakingBridge.sol:194]: Okok, well, we've usually assigned it a different criteria since it might be desirable to subsidize for example only deposit. For this reason please simplify it and make it return different ones.

File: src/bridges/convex/ConvexStakingBridge.sol (hunk 24 above)
[the remark this answers is not in the document]
Comment (benesjan) [reply src/bridges/convex/ConvexStakingBridge.sol:124]: Ok, you can solve that by moving the call up in the function.

[the remark this answers is not in the document]
Comment (VojtaKai) [reply src/bridges/convex/ConvexStakingBridge.sol:124]: By removing inputDeployedClone and outputDeployedClone, I could place the subsidy claim function back inside the convert function.

File: src/bridges/convex/ConvexStakingBridge.sol [hunk 25]
```diff
@@ -0,0 +1,265 @@
...
+import {ICurveRewards} from "../../interfaces/convex/ICurveRewards.sol";
+import {IRepConvexToken} from "../../interfaces/convex/IRepConvexToken.sol";
+
+/**
+ @notice A DefiBridge that allows user to stake Curve LP tokens through Convex Finance and earn boosted CRV 
+ without locking them in for an extended period of time. Plus earning CVX and possibly other rewards. 
+ @notice Staked tokens can be withdrawn (unstaked) any time.
+ @dev Convex Finance mints pool specific Convex LP token, however, not for the staking user (the bridge) directly.
+ RCT ERC20 token is deployed for each loaded pool and mirrors balance of minted Convex LP tokens for the bridge.
+ The difference is that RCT is owned directly by the bridge whereas Convex LP tokens are not and cannot be transferred to the Rollup Processor.
```
[the remark this answers is not in the document]
Comment (VojtaKai) [reply src/bridges/convex/ConvexStakingBridge.sol]: By 120 chars do you mean the max length of a row or a length of description for each @ spec?

[the remark this answers is not in the document]
Comment (VojtaKai) [reply src/bridges/convex/ConvexStakingBridge.sol]: The netstats will most likely need some more care and I'll try to make them more brief than they are now.
@dev should be only one or can I add multiple?

File: src/bridges/convex/ConvexStakingBridge.sol (hunk 24 above)
[the remark this answers is not in the document]
Comment (VojtaKai) [reply src/bridges/convex/ConvexStakingBridge.sol]: I am not sure I follow how moving the call up in the function would affect the total number of local variables?🤔 Call you please give me a more detailed explanation?

[the remark this answers is not in the document]
Comment (benesjan) [reply src/bridges/convex/ConvexStakingBridge.sol]: It would affect the number of variables on the stack at the time of execution of the code on that particular line. The limitation here is that EVM can only access variables on the stack up to 16 levels deep. Vyper language doesn't have this error since they instead allocate everything to memory instead of using stack. See [this article](https://medium.com/coinmonks/stack-too-deep-error-in-solidity-608d1bd6a1ea)

File: src/bridges/convex/ConvexStakingBridge.sol (hunk 25 above)
[the remark this answers is not in the document]
Comment (benesjan) [reply src/bridges/convex/ConvexStakingBridge.sol]: There can be multiple `@dev` tags. See [other bridges](https://github.com/AztecProtocol/aztec-connect-bridges/blob/master/src/bridges/liquity/TroveBridge.sol) for how to do it.

File: src/test/bridges/convex/ConvexStakingBridge.t.sol [hunk 26]
```diff
@@ -0,0 +1,518 @@
...
+    ) public {
+        vm.assume(_withdrawalAmount > 4012207);
+
+        uint256 poolLength = IConvexBooster(BOOSTER).poolLength();
+
+        bool testedPoolOpen;
+
+        _setupTestPoolIds(poolLength, _poolId1, _poolId2, _poolId3, _poolId4, 4); // Pool id 4 intentionally selected -> to prevent possibility that all selected pools yield zero CRV and CVX rewards in the specified time frame.
+
+        for (uint256 i = 0; i < poolIdsToTest.length; i++) {
```
Comment (benesjan) [inline src/test/bridges/convex/ConvexStakingBridge.t.sol:303]: Why are you iterating through all the pools and then skipping some? I would simply have 1 poolId as an input from the fuzzer and test it on that 1 pool. Default configuration of Foundry is running a 256 fuzz runs on each test runs so testing just 1 pool here should be ok.

**Review by benesjan — CHANGES_REQUESTED** [review]

File: src/interfaces/convex/ICurveLpToken.sol [hunk 27]
```diff
@@ -0,0 +1,8 @@
+// SPDX-License-Identifier: GPLv2
+pragma solidity >=0.8.4;
+
+interface ICurveLpToken {
```
[the remark this answers is not in the document]
Comment (VojtaKai) [reply src/interfaces/convex/ICurveLpToken.sol:4]: Is there any particular reason to not have an interface? I do understand that the interface is a "subset" of IERC20.

File: src/test/bridges/convex/ConvexStakingBridge.t.sol (hunk 26 above)
Comment (VojtaKai) [reply src/test/bridges/convex/ConvexStakingBridge.t.sol:303]: I do not iterate over all the pools, however, I selected 5 random pools from all the pools and these 5 pool Ids are then saved in poolIdsToTest and then it runs test for all five selected poolIds. The skipping mechanism is necessary to prevent tests to fails. (some pools just fail because they are not fully implemented, some pools can be close and that would fail the test, same pool cannot be tested within the same test twice, sometimes I need to add my own pool to make sure the tested assertion passes - e.g. here the pool id 4)
Do you mean that each fuzz parameters is changed 256 within the test? I read that somewhere but it didn't look like it from the test log (-vvvv)
Should I then use only a single fuzz poolId instead of five of them?

File: src/specs/convex/readme.md [hunk 28]
```diff
@@ -0,0 +1,89 @@
...
+
+Claimed rewards are the boosted CRV tokens, CVX tokens and for some pools some additional rewards. These rewards are minted for the bridge and are not recovered by the Rollup Processor.
```
[the remark this answers is not in the document]
Comment (VojtaKai) [reply src/specs/convex/readme.md:51]: There can be 0 to many rewards. Typically rewards are CVX token and CRV token. Some pools can have some extra rewards. Once I ran a test and only like 1/6 - 1/5 of all bridges had some extra rewards. Highest number I found was 3 if I remember correctly. I did try to figure out how to claim these extra rewards but I got lost in the process. It became surprisingly complicated and they also were some no name tokens.

[the remark this answers is not in the document]
Comment (benesjan) [reply src/specs/convex/readme.md:51]: Ok, please research which rewards it would even make sense to claim. If those extra rewards have tiny APR then it doesn't make sense to deal with that. But the CRV and CVX rewards should always get to the user. If the user doesn't get rewards out of the bridge then it's not very useful. When I checked matcha.xyz and 1inch they use curve to swap cvx and crv. For this reason I would dump both of these tokens to ETH and return them to the users as `outputAssetB`. Check out out curvesteth bridge for how to use curve to swap. It's pretty straightforward.
image

File: src/interfaces/convex/ICurveLpToken.sol (hunk 27 above)
[the remark this answers is not in the document]
Comment (benesjan) [reply src/interfaces/convex/ICurveLpToken.sol]: The token is an ERC20 token and everyone coding smart contracts knows what an ERC20 token is. Using custom interface name here is like renaming a standard string class to something nobody knows for no particular reason.

File: src/test/bridges/convex/ConvexStakingBridge.t.sol (hunk 26 above)
Comment (VojtaKai) [reply src/test/bridges/convex/ConvexStakingBridge.t.sol:303]: I am not sure if this is a good concept. I will essentially need to limit the generated poolId in order to bound it to the possible poolId range and check if the pool will not fail. Run the check in a while loop to guarantee that the selected pool will really pass the checks. Also the generated poolId has to be always greater than the number of available pools (vm.assume) otherwise bound could get stuck on the single failing poolId (last time I checked the bound function, if the number fit in the given range, the number was accepted and not changed anymore.). I sure can do all this and make sure that _getPoolId() will always return a valid pool but I think this is a lot of work when instead I could just skip some of the 256 tested pools.
Also for the withdraw test to check whether the rewards were claimed, a single poolId will not be enough, I will need the poolId 4 as well.
I'll brain a little more about this but today I didn't find any better solution for the _getPoolId()

File: src/bridges/convex/ConvexStakingBridge.sol [hunk 29]
```diff
@@ -0,0 +1,286 @@
...
+        AztecTypes.AztecAsset memory _outputAssetA,
+        uint256 _totalInputValue,
+        uint256 _interactionNonce
+    ) internal returns (uint256 outputValueA, uint256 outputValueB) {
+        PoolInfo memory selectedPool = pools[_outputAssetA.erc20Address];
+
+        uint256 startCurveLpTokens = IERC20(_outputAssetA.erc20Address).balanceOf(address(this));
+
+        // transfer ConvexLP tokens from CrvRewards back to the bridge
+        ICurveRewards(selectedPool.curveRewards).withdraw(_totalInputValue, true); // always claim rewards
```
Comment (benesjan) [inline src/bridges/convex/ConvexStakingBridge.sol:239]: This claims the rewards for the whole bridge, right? If that is the case then there is a bit of an issue with to whom you are returning the rewards. You are essentially returning the rewards of all the users to the users of the current round. If that is the case then it's a tricky situation and accounting might get complicated.

File: src/test/bridges/convex/ConvexStakingBridgeE2E.t.sol [hunk 30]
```diff
@@ -0,0 +1,243 @@
...
+        vm.label(staker, "Staker Contract Address");
+        vm.label(BENEFICIARY, "Beneficiary");
+    }
+
+    /**
+     * @notice Tests staking and withdrawing by constructing bridgeCallData and passing it directly
+     * to RollupProcessor function that initializes the interaction with the bridge.
+     * @dev Tests a randomly selected pool from all available pools.
+     */
+    function testStakeWithdrawFlow(uint64 _depositAmount, uint16 _poolId) public {
```
Comment (benesjan) [inline src/test/bridges/convex/ConvexStakingBridgeE2E.t.sol:58]: You should test that the accumulation of the rewards works. Are the CRV and CVX rewards distributed based on time? If yes, modifying block time should be sufficient for that.

**Review by benesjan — CHANGES_REQUESTED** [review]

File: yarn.lock [hunk 31]
```diff
@@ -16,7 +16,7 @@
 
 "@babel/highlight@^7.18.6":
   version "7.18.6"
```
[the remark this answers is not in the document]
Comment (VojtaKai) [reply yarn.lock:18]: It shouldn't. I'll checkout the file from master and commit that one.

File: src/test/bridges/convex/ConvexStakingBridgeE2E.t.sol [hunk 32]
```diff
@@ -0,0 +1,243 @@
...
+    address private rctImplementation;
+    address private rctClone;
+    address private staker;
+    address private gauge;
+    address private stash;
+    address private crvRewards;
+    // The reference to the convex staking bridge
+    ConvexStakingBridge private bridge;
+
+    uint16[] public invalidPids = [48]; // define invalid pids
```
[the remark this answers is not in the document]
Comment (VojtaKai) [reply src/test/bridges/convex/ConvexStakingBridgeE2E.t.sol:30]: Until now I only found this one but there could be more in the future.

File: src/bridges/convex/ConvexStakingBridge.sol (hunk 29 above)
Comment (VojtaKai) [reply src/bridges/convex/ConvexStakingBridge.sol:239]: Correct, it claims rewards for the whole bridge each time the withdraw function is called. Should we introduce some accounting management and proportional pay off? This would most likely turn the bridge into a stateful bridge

File: src/test/bridges/convex/ConvexStakingBridgeE2E.t.sol (hunk 30 above)
Comment (VojtaKai) [reply src/test/bridges/convex/ConvexStakingBridgeE2E.t.sol:58]: Well, it is going to be a bit trickier because CRV and CVX rewards are based on the stacked amount, performance of the pool (APY) and time. Also in some cases rewards can just be claimed but not exchanged for ether (because just too low amount of rewards was accumulated).
Also it was already tested in the unit tests and here I just check that the rewards claimed can be zero or greater.

Comment (VojtaKai) [reply src/test/bridges/convex/ConvexStakingBridgeE2E.t.sol:58]: So I will most likely need to tweek it to make it work

File: src/bridges/convex/ConvexStakingBridge.sol (hunk 29 above)
Comment (benesjan) [reply src/bridges/convex/ConvexStakingBridge.sol:239]: Yes, the user should receive only his rewards and not the rewards of the whole bridge.

File: src/test/bridges/convex/ConvexStakingBridgeE2E.t.sol (hunk 30 above)
Comment (benesjan) [reply src/test/bridges/convex/ConvexStakingBridgeE2E.t.sol:58]: If too low an amount of rewards is to be exchanged then you should not do the exchange at all since it would be a waste of gas. I would first implement the rewards accounting and then I would test if it works as expected. You could do that by manually minting the rewards to the bridge.

**Review by benesjan — CHANGES_REQUESTED** [review]
Please gas profile it and submit the values in the PR comment

File: src/bridges/convex/ConvexStakingBridge.sol [hunk 33]
```diff
@@ -0,0 +1,484 @@
...
+        } else {
+            revert ErrorLib.InvalidInput(); // invalid address or pool has not been loaded yet / RCT token not deployed yet
+        }
+
+        // Pays out subsidy to the rollupBeneficiary
+        SUBSIDY.claimSubsidy(
+            _computeCriteria(_inputAssetA.erc20Address, _outputAssetA.erc20Address), _rollupBeneficiary
+        );
+
+        emit ShowGas(startGas - gasleft());
```
[the remark this answers is not in the document]
Comment (VojtaKai) [reply src/bridges/convex/ConvexStakingBridge.sol:188]: Gas profiles
1. First deposit, first withdrawal, min deposited value 4e22
image

2. First deposit, first withdrawal, min deposited value 1
image

3. First and second deposit (days apart), first withdrawal, min deposited value 
image
